File: src/cdp_browser/actions.py
"""High-level browser actions via CDP."""
import time

from .client import (
    close_tab,
    execute_js,
    get_page_text,
    navigate,
    new_tab,
    wait_for_load,
)

# ---------------------------------------------------------------
# Web search


def web_search(query: str) -> list[dict]:
    """Search Google and return top results.

    Returns list of dicts: title, url, snippet.
    """
    import urllib.parse
    search_url = (
        "https://www.google.com/search?q="
        + urllib.parse.quote_plus(query)
    )
    tab = new_tab(search_url)
    try:
        time.sleep(2)  # let JS render results

        script = """
        (function() {
            var results = [];
            // Google 2025+ DOM: h3 inside .zReHs containers
            var headings = document.querySelectorAll('h3');
            for (var i = 0; i < headings.length && results.length < 10; i++) {
                var h = headings[i];
                // Walk up to find the nearest anchor
                var a = h.closest('a') ||
                        h.querySelector('a') ||
                        h.parentElement && h.parentElement.closest('a');
                if (!a || !a.href || a.href.startsWith('https://www.google')) continue;
                // Snippet: next sibling text block
                var container = h.closest('[data-sokoban-container]') ||
                                h.closest('.MjjYud') ||
                                h.parentElement;
                var snippetEl = container
                    ? container.querySelector('[data-sncf="1"]') ||
                      container.querySelector('.VwiC3b') ||
                      container.querySelector('span[style]')
                    : null;
                results.push({
                    title: h.textContent.trim(),
                    url: a.href,
                    snippet: snippetEl ? snippetEl.textContent.trim() : ''
                });
            }
            return JSON.stringify(results);
        })()
        """
        raw = execute_js(tab, script)
        import json
        results = json.loads(raw) if raw else []
        return [r for r in results if r.get("title")]
    finally:
        close_tab(tab)


# ---------------------------------------------------------------
# Web fetch


def web_fetch(url: str, max_chars: int = 10000) -> str:
    """Fetch a URL and return its main text content.

    Extracts article/main content, strips noise,
    truncates to max_chars.
    """
    tab = new_tab(url)
    try:
        time.sleep(2)  # let JS-heavy pages render

        script = """
        (function() {
            // Try semantic content containers first
            var el = document.querySelector('article') ||
                     document.querySelector('main') ||
                     document.querySelector('[role="main"]') ||
                     document.querySelector('.content') ||
                     document.querySelector('#content') ||
                     document.body;
            if (!el) return document.body.innerText;

            // Remove noisy elements
            var noisy = el.querySelectorAll(
                'nav, footer, header, aside, ' +
                '.nav, .footer, .header, .sidebar, ' +
                '.advertisement, .ad, script, style'
            );
            noisy.forEach(function(n) { n.remove(); });

            return el.innerText || el.textContent || '';
        })()
        """
        text = execute_js(tab, script) or ""
        text = text.strip()
        if len(text) > max_chars:
            text = text[:max_chars] + "\n...[truncated]"
        return text
    finally:
        close_tab(tab)


# ---------------------------------------------------------------
# Web screenshot


def web_screenshot(url: str) -> bytes:
    """Navigate to url and take a PNG screenshot."""
    from .client import take_screenshot
    tab = new_tab(url)
    try:
        time.sleep(2)
        return take_screenshot(tab)
    finally:
        close_tab(tab)


# ---------------------------------------------------------------
# LinkedIn


def linkedin_feed() -> list[dict]:
    """Read LinkedIn feed and return recent posts.

    Returns list of dicts: author, text, likes.
    """
    tab = new_tab("https://www.linkedin.com/feed/")
    try:
        time.sleep(4)  # feed takes a few seconds

        # Scroll down to load more posts
        execute_js(tab, "window.scrollBy(0, 800)")
        time.sleep(1)

        script = """
        (function() {
            var posts = [];
            var items = document.querySelectorAll(
                '.feed-shared-update-v2'
            );
            for (var i = 0; i < items.length; i++) {
                var el = items[i];
                var authorEl = el.querySelector(
                    '.update-components-actor__name'
                );
                var textEl = el.querySelector(
                    '.feed-shared-text'
                ) || el.querySelector(
                    '.update-components-text'
                );
                var likesEl = el.querySelector(
                    '.social-details-social-counts'
                );
                if (authorEl) {
                    posts.push({
                        author: authorEl.textContent.trim(),
                        text: textEl
                            ? textEl.textContent.trim()
                            : '',
                        likes: likesEl
                            ? likesEl.textContent.trim()
                            : ''
                    });
                }
            }
            return JSON.stringify(posts);
        })()
        """
        raw = execute_js(tab, script)
        import json
        posts = json.loads(raw) if raw else []
        return [p for p in posts if p.get("author")]
    finally:
        close_tab(tab)


def linkedin_like(
    post_index: int = 0, dry_run: bool = True
) -> dict:
    """Like a LinkedIn feed post by index.

    If dry_run=True, returns post info without clicking.
    """
    tab = new_tab("https://www.linkedin.com/feed/")
    try:
        time.sleep(4)

        # Get posts info
        script = """
        (function() {
            var items = document.querySelectorAll(
                '.feed-shared-update-v2'
            );
            var el = items[%d];
            if (!el) return JSON.stringify({error: 'post not found'});
            var authorEl = el.querySelector(
                '.update-components-actor__name'
            );
            var textEl = el.querySelector('.feed-shared-text') ||
                         el.querySelector('.update-components-text');
            var likeBtn = el.querySelector(
                'button[aria-label*="Like"]'
            ) || el.querySelector(
                '.reactions-react-button'
            );
            return JSON.stringify({
                author: authorEl
                    ? authorEl.textContent.trim() : '',
                text: textEl
                    ? textEl.textContent.trim().slice(0, 200) : '',
                like_button_found: likeBtn !== null
            });
        })()
        """ % post_index

        import json
        raw = execute_js(tab, script)
        info = json.loads(raw) if raw else {}

        if dry_run:
            info["dry_run"] = True
            info["action"] = "not liked (dry run)"
            return info

        # Click the like button
        click_script = """
        (function() {
            var items = document.querySelectorAll(
                '.feed-shared-update-v2'
            );
            var el = items[%d];
            if (!el) return 'post not found';
            var btn = el.querySelector(
                'button[aria-label*="Like"]'
            ) || el.querySelector(
                '.reactions-react-button'
            );
            if (!btn) return 'like button not found';
            btn.click();
            return 'clicked';
        })()
        """ % post_index

        result = execute_js(tab, click_script)
        time.sleep(1)
        info["action"] = result
        return info
    finally:
        close_tab(tab)

File: server_auto/src/cdp_browser/actions.rs
/// High-level browser actions via CDP.
use anyhow::Result;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use super::client::{close_tab, open_session};

// ---------------------------------------------------------------

#[derive(Debug, Serialize, Deserialize)]
pub struct SearchResult {
    pub title: String,
    pub url: String,
    pub snippet: String,
}

/// Search Google and return top results.
pub async fn web_search(query: &str) -> Result<Vec<SearchResult>> {
    let encoded = urlencoding::encode(query);
    let search_url = format!("https://www.google.com/search?q={encoded}");
    let (tab_id, mut session) = open_session(&search_url).await?;

    tokio::time::sleep(std::time::Duration::from_secs(2)).await;

    let script = r#"
    (function() {
        var results = [];
        var headings = document.querySelectorAll('h3');
        for (var i = 0; i < headings.length && results.length < 10; i++) {
            var h = headings[i];
            var a = h.closest('a') || h.querySelector('a') ||
                    (h.parentElement && h.parentElement.closest('a'));
            if (!a || !a.href || a.href.startsWith('https://www.google')) continue;
            var container = h.closest('[data-sokoban-container]') ||
                            h.closest('.MjjYud') || h.parentElement;
            var snippetEl = container ? (
                container.querySelector('[data-sncf="1"]') ||
                container.querySelector('.VwiC3b') ||
                container.querySelector('span[style]')
            ) : null;
            results.push({
                title: h.textContent.trim(),
                url: a.href,
                snippet: snippetEl ? snippetEl.textContent.trim() : ''
            });
        }
        return JSON.stringify(results);
    })()"#;

    let raw = session.evaluate(script).await?;
    let _ = close_tab(&tab_id).await;
    let results: Vec<SearchResult> = serde_json::from_str(raw.as_str().unwrap_or("[]"))
        .unwrap_or_default();
    Ok(results.into_iter().filter(|r| !r.title.is_empty()).collect())
}

/// Fetch a URL and return its main text content.
pub async fn web_fetch(url: &str, max_chars: usize) -> Result<String> {
    let (tab_id, mut session) = open_session(url).await?;
    tokio::time::sleep(std::time::Duration::from_secs(2)).await;

    let script = r#"
    (function() {
        var el = document.querySelector('article') ||
                 document.querySelector('main') ||
                 document.querySelector('[role="main"]') ||
                 document.querySelector('.content') ||
                 document.querySelector('#content') ||
                 document.body;
        if (!el) return document.body.innerText;
        var noisy = el.querySelectorAll(
            'nav, footer, header, aside, .nav, .footer, .header, .sidebar, .advertisement, .ad, script, style'
        );
        noisy.forEach(function(n) { n.remove(); });
        return el.innerText || el.textContent || '';
    })()"#;

    let raw = session.evaluate(script).await?;
    let _ = close_tab(&tab_id).await;
    let mut text = raw.as_str().unwrap_or("").trim().to_string();
    if text.len() > max_chars {
        text.truncate(max_chars);
        text.push_str("\n...[truncated]");
    }
    Ok(text)
}

/// Take a screenshot of a URL and return PNG bytes.
pub async fn web_screenshot(url: &str) -> Result<Vec<u8>> {
    let (tab_id, mut session) = open_session(url).await?;
    tokio::time::sleep(std::time::Duration::from_secs(2)).await;
    let png = session.screenshot().await?;
    let _ = close_tab(&tab_id).await;
    Ok(png)
}

// ---------------------------------------------------------------
// LinkedIn

#[derive(Debug, Serialize, Deserialize)]
pub struct LinkedInPost {
    pub author: String,
    pub text: String,
    pub likes: String,
}

pub async fn linkedin_feed() -> Result<Vec<LinkedInPost>> {
    let (tab_id, mut session) = open_session("https://www.linkedin.com/feed/").await?;
    tokio::time::sleep(std::time::Duration::from_secs(4)).await;
    session.evaluate("window.scrollBy(0, 800)").await?;
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;

    let script = r#"
    (function() {
        var posts = [];
        var items = document.querySelectorAll('.feed-shared-update-v2');
        for (var i = 0; i < items.length; i++) {
            var el = items[i];
            var authorEl = el.querySelector('.update-components-actor__name') ||
                           el.querySelector('.update-components-actor__title') ||
                           el.querySelector('[class*="actor__name"]');
            var textEl = el.querySelector('.feed-shared-text') ||
                         el.querySelector('.update-components-text') ||
                         el.querySelector('.attributed-text-segment-list__content');
            var likesEl = el.querySelector('.social-details-social-counts');
            if (authorEl) {
                posts.push({
                    author: (function(t) {
                        // Remove duplicate names (e.g. "Jim TarantinoJim Tarantino")
                        t = t.trim().split('\n')[0].trim();
                        var half = Math.floor(t.length / 2);
                        if (half > 2 && t.slice(0, half) === t.slice(half)) t = t.slice(0, half);
                        return t;
                    })(authorEl.textContent),
                    text: textEl ? textEl.textContent.trim() : '',
                    likes: likesEl ? likesEl.textContent.trim() : ''
                });
            }
        }
        return JSON.stringify(posts);
    })()"#;

    let raw = session.evaluate(script).await?;
    let _ = close_tab(&tab_id).await;
    let posts: Vec<LinkedInPost> = serde_json::from_str(raw.as_str().unwrap_or("[]"))
        .unwrap_or_default();
    Ok(posts.into_iter().filter(|p| !p.author.is_empty()).collect())
}

pub async fn linkedin_like(post_index: usize, dry_run: bool) -> Result<Value> {
    let (tab_id, mut session) = open_session("https://www.linkedin.com/feed/").await?;
    tokio::time::sleep(std::time::Duration::from_secs(4)).await;

    let info_script = format!(r#"
    (function() {{
        var items = document.querySelectorAll('.feed-shared-update-v2');
        var el = items[{post_index}];
        if (!el) return JSON.stringify({{error: 'post not found'}});
        var authorEl = el.querySelector('.update-components-actor__name') ||
                       el.querySelector('.update-components-actor__title') ||
                       el.querySelector('[class*="actor__name"]');
        var textEl = el.querySelector('.feed-shared-text') ||
                     el.querySelector('.update-components-text') ||
                     el.querySelector('.attributed-text-segment-list__content');
        var likeBtn = el.querySelector('button[aria-label*="Like"]') || el.querySelector('.reactions-react-button');
        var authorRaw = authorEl ? authorEl.textContent.trim().split('\n')[0].trim() : '';
        var half = Math.floor(authorRaw.length / 2);
        if (half > 2 && authorRaw.slice(0, half) === authorRaw.slice(half)) authorRaw = authorRaw.slice(0, half);
        return JSON.stringify({{
            author: authorRaw,
            text: textEl ? textEl.textContent.trim().slice(0, 200) : '',
            like_button_found: likeBtn !== null
        }});
    }})()"#);

    let raw = session.evaluate(&info_script).await?;
    let mut info: Value = serde_json::from_str(raw.as_str().unwrap_or("{}")).unwrap_or_default();

    if dry_run {
        info["dry_run"] = json!(true);
        info["action"] = json!("not liked (dry run)");
        let _ = close_tab(&tab_id).await;
        return Ok(info);
    }

    let click_script = format!(r#"
    (function() {{
        var items = document.querySelectorAll('.feed-shared-update-v2');
        var el = items[{post_index}];
        if (!el) return 'post not found';
        var btn = el.querySelector('button[aria-label*="Like"]') || el.querySelector('.reactions-react-button');
        if (!btn) return 'like button not found';
        btn.click();
        return 'clicked';
    }})()"#);

    let result = session.evaluate(&click_script).await?;
    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    info["action"] = result;
    let _ = close_tab(&tab_id).await;
    Ok(info)
}
